File: content-solutions/queue-deque/deque-steady-signal.java
import java.util.*;

class Solution {
    public int solution(int[] signal, int limit) {
        Deque<Integer> maxq = new ArrayDeque<>();
        Deque<Integer> minq = new ArrayDeque<>();
        int left = 0, best = 0;
        for (int right = 0; right < signal.length; right++) {
            int value = signal[right];
            while (!maxq.isEmpty() && signal[maxq.peekLast()] <= value) maxq.pollLast();
            maxq.offerLast(right);
            while (!minq.isEmpty() && signal[minq.peekLast()] >= value) minq.pollLast();
            minq.offerLast(right);
            while (signal[maxq.peekFirst()] - signal[minq.peekFirst()] > limit) {
                left++;
                if (maxq.peekFirst() < left) maxq.pollFirst();
                if (minq.peekFirst() < left) minq.pollFirst();
            }
            best = Math.max(best, right - left + 1);
        }
        return best;
    }
}

File: content-solutions/queue-deque/deque-steady-signal.js
function solution(signal, limit) {
  const n = signal.length;
  const maxq = new Array(n);
  const minq = new Array(n);
  let maxHead = 0,
    maxTail = 0,
    minHead = 0,
    minTail = 0;
  let left = 0;
  let best = 0;
  for (let right = 0; right < n; right += 1) {
    const value = signal[right];
    while (maxTail > maxHead && signal[maxq[maxTail - 1]] <= value) maxTail -= 1;
    maxq[maxTail++] = right;
    while (minTail > minHead && signal[minq[minTail - 1]] >= value) minTail -= 1;
    minq[minTail++] = right;
    while (signal[maxq[maxHead]] - signal[minq[minHead]] > limit) {
      left += 1;
      if (maxq[maxHead] < left) maxHead += 1;
      if (minq[minHead] < left) minHead += 1;
    }
    best = Math.max(best, right - left + 1);
  }
  return best;
}
